### PR TITLE
[ISSUE-306][IMPROVEMENT] Handle change partition request in batch

### PR DESCRIPTION
**Pull request by AngersZhuuuu**
### Related issues.
https://github.com/alibaba/RemoteShuffleService/issues/306

### How was this patch tested?

/cc @related-reviewer

/assign @main-reviewer

### CODE REVIEW

**Comment by AngersZhuuuu** [comment]
ping @waitinfuture @FMX @pan3793

File: client/src/main/scala/com/aliyun/emr/rss/client/write/LifecycleManager.scala [hunk 1]
```diff
@@ -1144,21 +1226,19 @@ class LifecycleManager(appId: String, val conf: RssConf) extends RpcEndpoint wit
...
+  /**
+   * Reallocate slots from available workers.
+   *
+   * @param partitionIdToEpochs The list of origin partition's id and epoch number.
+   * @param candidates Available worker.
+   * @return Reallocated PartitionLocations.
+   */
   private def reallocateSlotsFromCandidates(
-      oldPartitions: List[PartitionLocation],
+      partitionIdToEpochs: Array[(Int, Int)],
```
Comment (pan3793) [inline client/src/main/scala/com/aliyun/emr/rss/client/write/LifecycleManager.scala:1237]: unnecessary change? but it's good to use array or scala collection so we can avoid converting.

**Review by pan3793 — APPROVED** [review]

File: client/src/main/scala/com/aliyun/emr/rss/client/write/LifecycleManager.scala (hunk 1 above)
Comment (AngersZhuuuu) [reply client/src/main/scala/com/aliyun/emr/rss/client/write/LifecycleManager.scala:1237]: > unnecessary change? but it's good to use array or scala collection so we can avoid converting.

Double check code, seems we don't need 
```
   val partitionIdToEpochs = changePartitions.map { split =>
      if (split.oldPartition != null) {
        split.oldPartition.getId -> split.oldPartition.getEpoch
      } else {
        split.partitionId -> split.epoch
      }
    }
```

This change is unnecessary

File: client/src/main/scala/com/aliyun/emr/rss/client/write/LifecycleManager.scala [hunk 2]
```diff
@@ -499,62 +563,72 @@ class LifecycleManager(appId: String, val conf: RssConf) extends RpcEndpoint wit
...
+      // reply the master location of this partition.
+      val newMasterLocation =
+        if (masterLocations != null && masterLocations.size() > 0) {
+          masterLocations.asScala.head
+        } else {
+          slaveLocations.asScala.head.getPeer
+        }
+      reply(newMasterLocation.getId, ChangeLocationResponse(StatusCode.SUCCESS, newMasterLocation))
+      logDebug(s"Renew $shuffleId ${newMasterLocation.getId}" +
+        s"${newMasterLocation.getEpoch - 1} -> ${newMasterLocation.getEpoch} partition success.")
```
Comment (pan3793) [inline client/src/main/scala/com/aliyun/emr/rss/client/write/LifecycleManager.scala:630]: space

File: client/src/main/scala/com/aliyun/emr/rss/client/write/LifecycleManager.scala [hunk 3]
```diff
@@ -1163,6 +1238,7 @@ class LifecycleManager(appId: String, val conf: RssConf) extends RpcEndpoint wit
     slots
   }
 
+
```
Comment (pan3793) [inline client/src/main/scala/com/aliyun/emr/rss/client/write/LifecycleManager.scala:1241]: unnecessary line

File: client/src/main/scala/com/aliyun/emr/rss/client/write/LifecycleManager.scala (hunk 2 above)
Comment (AngersZhuuuu) [reply client/src/main/scala/com/aliyun/emr/rss/client/write/LifecycleManager.scala]: Done

File: client/src/main/scala/com/aliyun/emr/rss/client/write/LifecycleManager.scala [hunk 4]
```diff
@@ -499,62 +563,84 @@ class LifecycleManager(appId: String, val conf: RssConf) extends RpcEndpoint wit
...
+    // remove together to reduce lock time
+    def replyFailure(response: ChangeLocationResponse): Unit = {
       requests.synchronized {
-        requests.remove(partitionId)
-      }.asScala.foreach(_.reply(response))
+        changePartitions.flatMap { changePartition =>
+          requests.remove(changePartition.partitionId).asScala.toList
+        }
+      }.foreach(_.context.reply(response))
     }
```
Comment (AngersZhuuuu) [inline client/src/main/scala/com/aliyun/emr/rss/client/write/LifecycleManager.scala:604]: Change a lot the the reply method to reduce frequent sync, pls take a look again. @pan3793

File: client/src/main/scala/com/aliyun/emr/rss/client/write/LifecycleManager.scala (hunk 2 above)
Comment (AngersZhuuuu) [reply client/src/main/scala/com/aliyun/emr/rss/client/write/LifecycleManager.scala]: done

File: common/src/main/scala/com/aliyun/emr/rss/common/RssConf.scala [hunk 5]
```diff
@@ -825,6 +825,14 @@ object RssConf extends Logging {
     conf.getSizeAsBytes("rss.partition.split.minimum.size", "1m")
   }
 
+  def changePartitionNumThreads(conf: RssConf): Int = {
+    conf.getInt("rss.change.partition.numThreads", 8)
```
[the remark this answers is not in the document]
Comment (AngersZhuuuu) [reply common/src/main/scala/com/aliyun/emr/rss/common/RssConf.scala:829]: Done

File: client/src/main/scala/com/aliyun/emr/rss/client/write/LifecycleManager.scala [hunk 6]
```diff
@@ -104,6 +113,14 @@ class LifecycleManager(appId: String, val conf: RssConf) extends RpcEndpoint wit
   private val responseCheckerThread =
     ThreadUtils.newDaemonSingleThreadScheduledExecutor("rss-master-resp-checker")
 
+  private val changePartitionExecutors = ThreadUtils.newDaemonCachedThreadPool(
+    "lifecycle-manager-change-partition",
```
Comment (pan3793) [inline client/src/main/scala/com/aliyun/emr/rss/client/write/LifecycleManager.scala:117]: since it lives in the Spark process, can we add an `rss-` prefix to the thread name?

File: client/src/main/scala/com/aliyun/emr/rss/client/write/LifecycleManager.scala [hunk 7]
```diff
@@ -104,6 +113,14 @@ class LifecycleManager(appId: String, val conf: RssConf) extends RpcEndpoint wit
...
+    RssConf.changePartitionNumThreads(conf))
+  private val handleChangePartitionRequestBatchInterval =
+    RssConf.handleChangePartitionRequestBatchInterval(conf)
+  private val changePartitionSchedulerThread =
+    ThreadUtils.newDaemonSingleThreadScheduledExecutor("partition-split-scheduler")
```
Comment (pan3793) [inline client/src/main/scala/com/aliyun/emr/rss/client/write/LifecycleManager.scala:122]: ditto, thread name

File: client/src/main/scala/com/aliyun/emr/rss/client/write/LifecycleManager.scala (hunk 6 above)
Comment (AngersZhuuuu) [reply client/src/main/scala/com/aliyun/emr/rss/client/write/LifecycleManager.scala]: > since it lives in the Spark process, can we add an `rss-` prefix to the thread name?

Yea, done

File: client/src/main/scala/com/aliyun/emr/rss/client/write/LifecycleManager.scala (hunk 7 above)
Comment (AngersZhuuuu) [reply client/src/main/scala/com/aliyun/emr/rss/client/write/LifecycleManager.scala]: > ditto, thread name

done

**Review by pan3793 — APPROVED** [review]

File: client/src/main/scala/com/aliyun/emr/rss/client/write/LifecycleManager.scala [hunk 8]
```diff
@@ -1144,12 +1231,13 @@ class LifecycleManager(appId: String, val conf: RssConf) extends RpcEndpoint wit
...
 
-  private def reallocateForNonExistPartitionLocationFromCandidates(
-      partitionId: Int,
-      oldEpochId: Int,
+  private def reallocateChangePartitionRequestSlotsFromCandidates(
+      oldPartitions: List[ChangePartitionRequest],
       candidates: List[WorkerInfo]): WorkerResource = {
     val slots = new WorkerResource()
-    allocateFromCandidates(partitionId, oldEpochId, candidates, slots)
+    oldPartitions.foreach { partition =>
```
Comment (waitinfuture) [inline client/src/main/scala/com/aliyun/emr/rss/client/write/LifecycleManager.scala:1238]: There are cases where oldPartition is null, see ShuffleClientImpl.pushOrMergeData
```angular2html
    // get location
    if (!map.containsKey(partitionId)
        && !revive(
            applicationId,
            shuffleId,
            mapId,
            attemptId,
            partitionId,
            0,
            null,
            StatusCode.PUSH_DATA_FAIL_NON_CRITICAL_CAUSE)) {
      throw new IOException(
          "Revive for shuffle " + shuffleKey + " partitionId " + partitionId + " failed.");
    }
```

File: client/src/main/scala/com/aliyun/emr/rss/client/write/LifecycleManager.scala [hunk 9]
```diff
@@ -149,6 +167,43 @@ class LifecycleManager(appId: String, val conf: RssConf) extends RpcEndpoint wit
...
+                        }.toArray
+                        batchHandleChangePartitions(
+                          distinctPartitions.head.applicationId,
+                          shuffleId,
+                          distinctPartitions)
+                      }
+                    }
+                  }
+                  // After put request into executor, clean current requests.
+                  requests.clear()
```
Comment (waitinfuture) [inline client/src/main/scala/com/aliyun/emr/rss/client/write/LifecycleManager.scala:193]: Should not clear() here, batchHandleChangePartitions will remove

File: client/src/main/scala/com/aliyun/emr/rss/client/write/LifecycleManager.scala [hunk 10]
```diff
@@ -499,62 +564,84 @@ class LifecycleManager(appId: String, val conf: RssConf) extends RpcEndpoint wit
...
+    // Blacklist all failed workers
+    if (changePartitions.exists(_.causes.isDefined)) {
+      changePartitions.filter(_.causes.isDefined).foreach { changePartition =>
+        blacklistPartition(shuffleId, changePartition.oldPartition, changePartition.causes.get)
+      }
+    }
+
+    // remove together to reduce lock time
+    def replySuccess(locations: Array[PartitionLocation]): Unit = {
+      requests.synchronized {
```
Comment (waitinfuture) [inline client/src/main/scala/com/aliyun/emr/rss/client/write/LifecycleManager.scala:589]: Better to rename requests to requestsMap

File: client/src/main/scala/com/aliyun/emr/rss/client/write/LifecycleManager.scala (hunk 9 above)
Comment (AngersZhuuuu) [reply client/src/main/scala/com/aliyun/emr/rss/client/write/LifecycleManager.scala:193]: > Should not clear() here, batchHandleChangePartitions will remove

Oh, yea,  first I use an external map to store change partition request then merge it in to origin map, remain this. Good catch.

File: client/src/main/scala/com/aliyun/emr/rss/client/write/LifecycleManager.scala (hunk 8 above)
Comment (AngersZhuuuu) [reply client/src/main/scala/com/aliyun/emr/rss/client/write/LifecycleManager.scala:1238]: > There are cases where oldPartition is null, see ShuffleClientImpl.pushOrMergeData
> 
> ```
>     // get location
>     if (!map.containsKey(partitionId)
>         && !revive(
>             applicationId,
>             shuffleId,
>             mapId,
>             attemptId,
>             partitionId,
>             0,
>             null,
>             StatusCode.PUSH_DATA_FAIL_NON_CRITICAL_CAUSE)) {
>       throw new IOException(
>           "Revive for shuffle " + shuffleKey + " partitionId " + partitionId + " failed.");
>     }
> ```

A mistake, here I forgot to change parameter name, should be `changePartitionRequests`

File: client/src/main/scala/com/aliyun/emr/rss/client/write/LifecycleManager.scala (hunk 10 above)
Comment (AngersZhuuuu) [reply client/src/main/scala/com/aliyun/emr/rss/client/write/LifecycleManager.scala:589]: Done

File: client/src/main/scala/com/aliyun/emr/rss/client/write/LifecycleManager.scala [hunk 11]
```diff
@@ -104,6 +113,15 @@ class LifecycleManager(appId: String, val conf: RssConf) extends RpcEndpoint wit
...
     ThreadUtils.newDaemonSingleThreadScheduledExecutor("rss-master-resp-checker")
 
+  private val changePartitionExecutors = ThreadUtils.newDaemonCachedThreadPool(
+    "rss-lifecycle-manager-change-partition-executor",
+    RssConf.changePartitionNumThreads(conf))
+  private val handleChangePartitionRequestBatchInterval =
+    RssConf.handleChangePartitionRequestBatchInterval(conf)
+  private val changePartitionSchedulerThread =
+    ThreadUtils.newDaemonSingleThreadScheduledExecutor(
+      "rss-lifecycle-maager-change-partition-scheduler")
```
[the remark this answers is not in the document]
Comment (AngersZhuuuu) [reply client/src/main/scala/com/aliyun/emr/rss/client/write/LifecycleManager.scala:123]: Done

File: client/src/main/scala/com/aliyun/emr/rss/client/write/LifecycleManager.scala [hunk 12]
```diff
@@ -149,6 +167,41 @@ class LifecycleManager(appId: String, val conf: RssConf) extends RpcEndpoint wit
       0,
       heartbeatIntervalMs,
       TimeUnit.MILLISECONDS)
+
+    changePartitionSchedulerThread.scheduleAtFixedRate(
+      new Runnable {
+        override def run(): Unit = {
+          try {
+            while (true) {
```
[the remark this answers is not in the document]
Comment (AngersZhuuuu) [reply client/src/main/scala/com/aliyun/emr/rss/client/write/LifecycleManager.scala]: Yea，done

File: client/src/main/scala/com/aliyun/emr/rss/client/write/LifecycleManager.scala [hunk 13]
```diff
@@ -499,62 +562,84 @@ class LifecycleManager(appId: String, val conf: RssConf) extends RpcEndpoint wit
...
-    val (masterLocations, slavePartitions) = newlyAllocatedLocation.asScala.head._2
-    // reply the master location of this partition.
-    val newMasterLocation =
-      if (masterLocations != null && masterLocations.size() > 0) {
-        masterLocations.asScala.head
-      } else {
-        slavePartitions.asScala.head.getPeer
-      }
+    val newMasterLocations =
+      newlyAllocatedLocations.asScala.map { case (workInfo, (masterLocations, slaveLocations)) =>
```
[the remark this answers is not in the document]
Comment (AngersZhuuuu) [reply client/src/main/scala/com/aliyun/emr/rss/client/write/LifecycleManager.scala]: > Should use flatMap and return all masterLocations

Updated

**Comment by waitinfuture** [comment]
This change impacts critical path, I think we'd better do a complete test involving revive and split to ensure correctness.

**Review by FMX — APPROVED** [review]
LGTM. Thanks.
